sql/opt_range.cc

```diff
@@ -1671,11 +1691,19 @@ int QUICK_RANGE_SELECT::init_ror_merged_scan(bool reuse_handler)
   if (reuse_handler)
   {
     DBUG_PRINT("info", ("Reusing handler %p", file));
+
+    head->column_bitmaps_set_no_signal(&column_bitmap, &column_bitmap);
```

Comment:
I might be missing something, but this call looks redundant to me: `mark_columns_used_by_index()` calls `TABLE::mark_columns_used_by_index` which makes this call
```
  column_bitmaps_set(bitmap, bitmap);
```
where `bitmap` points to TABLE::tmp_set.

Reply:
This is old code so I'm not 100% positive - based on my reading of the code it seems to be related to the earlier question regarding `no_keyread` - if no_keyread is true, then `prepare_position` call below might update column_bitmap (not tmp, if `mark_columns_used_by_index` isn't called):

```
  /* We don't have to set 'head->keyread' here as the 'file' is unique */
  if (!head->no_keyread)
    head->mark_columns_used_by_index(index);
  head->prepare_for_positon();
```

```
void TABLE::prepare_for_position()
{
  DBUG_ENTER("TABLE::prepare_for_position");

  if ((file->ha_table_flags() & HA_PRIMARY_KEY_REQUIRED_FOR_POSITION) &&
      s->primary_key < MAX_KEY)
  {
    mark_columns_used_by_index_no_reset(s->primary_key, read_set);
    /* signal change */
    file->column_bitmaps_signal();
  }
  DBUG_VOID_RETURN;
}
```

The whole thing seems rather needlessly convoluted IMO. I think the better option here is to call `mark_columns_used_by_index_no_reset` that modifies `column_bitmap` directly without relying on `tmp_set`, and it would also get rid of `bitmap_copy(&column_bitmap, head->read_set);` since all changes would've been made in `column_bitmap` directly, but there might be some subtle side effects that I may have missed so for now my preference is to keep it as-is (and also make future merges a bit easier when we port this to 8.0).